diff --git a/include/scheduler.h b/include/scheduler.h
index 99bfd36..ff6251e 100644
--- a/include/scheduler.h
+++ b/include/scheduler.h
@@ -145,6 +145,23 @@ public:
     FCFS(int p = NORMAL, Tn & ... an);
 };
 
+// !@#
+// Rate Monotonic
+class RM: public Priority
+{
+public:
+    static const bool preemptive = true;
+    static const bool timed = true;
+
+public:
+    // used for aperiodic threads
+    RM(int priority): Priority(priority) {}
+
+    // used for periodic threads
+    template <typename ... Tn>
+    RM(const Microsecond & period, int priority, Tn & ... an);
+};
